New entry go goroutines (#1410)
Codecademy/docs: content/go/concepts/goroutines/goroutines.md

---
Title: 'Goroutines'
Description: 'Goroutines are functions and methods that run concurrently with other functions and methods.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Control Flow'
  - 'Threads'
  - 'Concurrency'
  - 'Functions'
CatalogContent:
  - 'learn-go'
  - 'paths/back-end-engineer-career-path'
---

**Goroutines** are functions and methods that run concurrently with other functions and methods.

> Although goroutines share some similarities with threads, goroutines are not threads. Threads are different processes within a program that are running at the same time. Threads are created by the program; they are smaller and lighter processes and have their own flow of control. Running multiple processes simultaneously allows complex programs to run without holding up the execution of the rest of the program. Go supports multi-threading with goroutines. Goroutines are lightweight threads managed by the Go runtime.

Reviewer: ```suggestion
Although goroutines share some similarities with threads, there are important differences that include the following:

- Threads depend on the hardware of the host computer's operating system, whereas goroutines do not.
- Compared with threads, goroutines are cost-effective and use less resources to function.
- Goroutines do not use thread-local storage and, thus, do not have a unique ID.

Comparatively, goroutines are light-weight versions of threads that operate within the context of the Go runtime.
```

## Syntax

> 
> To start a goroutine, invoke a previously defined function or method with the `go` keyword:
> 

Reviewer: ```suggestion

```

```pseudo
func myFunction(parameter) returnType {

>    //body of the function

Reviewer: ```suggestion
   // function body
```

}

go myFunction()
```

> In the code above the `myFunction()` function is defined. Then the function is invoked with the `go` keyword. `myFunction()` will run concurrently with the calling function.
> 
> > **Note:** for any goroutines to run, the `main()` function should be running. When the `main()` function terminates, the program will be terminated, and no other goroutine will run.

Reviewer: ```suggestion
A goroutine is started by invoking a previously defined function or method, `myFunction()`, with the `go` keyword.

> **Note:** For any goroutines to run, the `main()` function must defined and executed. When the `main()` function terminates, the program will be terminated, and no other goroutine will run.
```

## Example

> In the example below, the `myGoroutine()` function is defined to print out the text `This is my first goroutine`. Then in the `main()` function, `myGoroutine()` is called as a goroutine with the `go` keyword. Afterwards, the `main()` function is set to wait two seconds with the `time.Sleep()` function and then print out the text `This is the main function`.

Reviewer: ```suggestion
In the following example, a goroutine is defined to print out text. Inside the `main()` function, `myGoroutine()` is called with the `go` keyword and causes a time delay of 2 seconds with `time.Sleep()` before printing the text:
```

```go

> func myGoroutine() {  

Reviewer: ```suggestion
package main
import (
  "fmt"
  "time"
)

func myGoroutine() {  
```

    fmt.Println("This is my first goroutine")
}

func main() {  

>     go myGoroutine()
>     time.Sleep(2 * time.Second)
>     fmt.Println("This is the main function")

Reviewer: ```suggestion
  go myGoroutine()
  time.Sleep(2 * time.Second)
  fmt.Println("This is the main function")
```

}
```

> This example results in the following output:

Reviewer: ```suggestion
This results in the following output:
```

```shell
This is my first goroutine
This is the main function
```

> **Note:** When a goroutine is started, the goroutine call returns immediately. The program does not wait for the goroutine to finish executing. After the goroutine call, the program continues the next line of code, and any return values from the goroutine are ignored. The `time.Sleep()` function keeps the `main()` function running while the `myGoroutine()` function executes. Without the two seconds waiting time in the `time.Sleep()` function, the second print statement could be immediately completed, and the `main()` function could terminate before the goroutine is completed.

## Codebyte Example

> The example below defines a function called `print()` to print out a string, which is first called as a goroutine and then as a function:

Reviewer: ```suggestion
The example below defines a `print()` function to print out a string, which is first called as a goroutine and then as a regular function:
```

---
Title: 'Goroutines'
Description: 'Goroutines are functions and methods that run concurrently with other functions and methods.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Control Flow'
  - 'Threads'
  - 'Concurrency'
  - 'Functions'
CatalogContent:
  - 'learn-go'
  - 'paths/back-end-engineer-career-path'
---

**Goroutines** are functions and methods that run concurrently with other functions and methods.

> ## The difference between threads and goroutines

Reviewer 2: I'd get rid of the header and incorporate the following paragraph into the introduction.

```go
func myGoroutine() {  
    fmt.Println("This is my first goroutine")
}

func main() {  
    go myGoroutine()
    time.Sleep(2 * time.Second)
    fmt.Println("This is the main function")
}
```

This example results in the following output:

```shell
This is my first goroutine
This is the main function
```

> **Note:** When a goroutine is started, the goroutine call returns immediately. The program does not wait for the goroutine to finish executing. After the goroutine call, the program continues the next line of code, and any return values from the goroutine are ignored. The `time.Sleep()` function keeps the `main()` function running while the `myGoroutine()` function executes. Without the two seconds waiting time in the `time.Sleep()` function, the second print statement could be immediately completed, and the `main()` function could terminate before the goroutine is completed.

## Codebyte Example

The example below defines a function called `print()` to print out a string, which is first called as a goroutine and then as a function:

```codebyte/golang
package main
import (
  "fmt"
  "time"
)

// define a function
func print(text string) {

> 

Reviewer 2: ```suggestion
```
